### PR TITLE
"mc admin config get/set" command support for partial configs

### LINKED ISSUE #2509: mc admin config `setkeys` command support
Setting config.json key values, by specifying single or multiple key/value pairs, is going to be supported with the new `mc admin config  setkeys` command.

See the [setkeys requirements document](https://gist.github.com/ebozduman/a26900558bb8ddb3d425134047fe86cf) for more detailed information.

### PR DESCRIPTION
**Pull request by ebozduman**
Fixes #2509.
Support changes for partial configurations for `mc admin config get` and `mc admin config set` commands.

### CODE REVIEW

File: cmd/admin-config-get.go [hunk 1]
```diff
@@ -67,31 +81,43 @@ func (u configGetMessage) JSON() string {
...
+	if isFullConfigGet {
+		// Call get config API
+		c, e := client.GetConfig()
+		fatalIf(probe.NewError(e), "Cannot get server configuration file.")
+		printMsg(configGetMessage{Config: string(c)})
+	} else {
+		argsList := ctx.Args().Tail()
+		// Call get config keys API
+		c, e := client.GetConfigKeys(argsList)
+		fatalIf(probe.NewError(e), "Cannot get server configuration file.")
```
Comment (kannappanr) [inline cmd/admin-config-get.go]: The message has to reflect the fact that we are requesting some keys and not a whole file.

**Comment by kannappanr** [comment]
overall looks good. Will test a bit more before approving.

**Comment by vadmeste** [comment]
@ebozduman can you rename "tmp" commit ? or squash your two commits

File: cmd/admin-config-set.go [hunk 2]
```diff
@@ -81,37 +88,55 @@ func (u configSetMessage) JSON() string {
 }
 
 // checkAdminConfigSetSyntax - validate all the passed arguments
-func checkAdminConfigSetSyntax(ctx *cli.Context) {
-	if len(ctx.Args()) == 0 || len(ctx.Args()) > 2 {
+func checkAdminConfigSetSyntax(ctx *cli.Context) string {
+	if len(ctx.Args()) == 0 {
 		cli.ShowCommandHelpAndExit(ctx, "set", 1) // last argument is exit code
 	}
+	if len(ctx.Args()) == 1 {
```
Comment (vadmeste) [inline cmd/admin-config-set.go]: I think we can move this to mainAdminConfigSet

File: cmd/admin-config-get.go [hunk 3]
```diff
@@ -67,31 +81,43 @@ func (u configGetMessage) JSON() string {
...
+	aliasedURL := ctx.Args().Get(0)
+	isFullConfigGet := false
 
-	checkAdminConfigGetSyntax(ctx)
-
-	// Get the alias parameter from cli
-	args := ctx.Args()
-	aliasedURL := args.Get(0)
+	// Check command arguments
+	if checkAdminConfigGetSyntax(ctx) == "fullGet" {
```
Comment (vadmeste) [inline cmd/admin-config-get.go]: Suggestion: let's check for command syntax the first thing in this function

File: cmd/admin-config-set.go [hunk 4]
```diff
@@ -81,37 +88,55 @@ func (u configSetMessage) JSON() string {
...
 
 // main config set function
 func mainAdminConfigSet(ctx *cli.Context) error {
+	// Initializations
+	aliasedURL := ctx.Args().Get(0)
+	isFullConfigSet := false
 
 	// Check command arguments
-	checkAdminConfigSetSyntax(ctx)
+	if checkAdminConfigSetSyntax(ctx) == "fullSet" {
```
Comment (vadmeste) [inline cmd/admin-config-set.go]: Suggestion: let's make the check the first thing to do in this function

File: cmd/admin-config-set.go [hunk 5]
```diff
@@ -81,37 +88,55 @@ func (u configSetMessage) JSON() string {
...
+	if isFullConfigSet {
+		// Call set config API
+		fatalIf(probe.NewError(client.SetConfig(os.Stdin)), "Cannot set server configuration file.")
+		// Print set config result
+		printMsg(configSetMessage{setConfigStatus: true})
+	} else {
+		argsMap := make(map[string]string)
+		for _, arg := range ctx.Args().Tail() {
+			argSplit := strings.SplitN(arg, "=", 2)
+			if strings.Index(arg, "{") == -1 && len(argSplit)%2 == 1 {
```
Comment (vadmeste) [inline cmd/admin-config-set.go]: Can you add a comment what are we doing this for ?

File: cmd/admin-config-get.go [hunk 6]
```diff
@@ -22,42 +22,56 @@ import (
...
-// String colorized service status message.
+// String returns config info as a string
 func (u configGetMessage) String() string {
-	return string(u.Config)
+	if len(u.argsList) == 0 {
+		return string(u.Config)
+	}
+	var str string
+	for _, key := range u.argsList {
+		val := gjson.Get(string(u.Config), key)
```
[the remark this answers is not in the document]
Comment (ebozduman) [reply cmd/admin-config-get.go]: OK.

File: cmd/admin-config-get.go [hunk 7]
```diff
@@ -67,31 +81,43 @@ func (u configGetMessage) JSON() string {
...
 
-// checkAdminConfigGetSyntax - validate all the passed arguments
-func checkAdminConfigGetSyntax(ctx *cli.Context) {
-	if len(ctx.Args()) == 0 || len(ctx.Args()) > 2 {
+// checkAdminConfigGetSyntax - validates arguments
+func checkAdminConfigGetSyntax(ctx *cli.Context) string {
+	if len(ctx.Args()) == 0 {
 		cli.ShowCommandHelpAndExit(ctx, "get", 1) // last argument is exit code
 	}
+	if len(ctx.Args()) == 1 {
```
[the remark this answers is not in the document]
Comment (ebozduman) [reply cmd/admin-config-get.go]: OK

File: cmd/admin-config-get.go (hunk 1 above)
Comment (ebozduman) [reply cmd/admin-config-get.go]: OK

File: cmd/admin-config-set.go (hunk 2 above)
Comment (ebozduman) [reply cmd/admin-config-set.go]: OK

File: cmd/admin-config-set.go (hunk 4 above)
Comment (ebozduman) [reply cmd/admin-config-set.go]: OK

File: cmd/admin-config-set.go (hunk 5 above)
Comment (ebozduman) [reply cmd/admin-config-set.go]: OK

**Comment by kannappanr** [comment]
We are re-designing the config backend. Marking this PR as blocked until then.

**Comment by kannappanr** [comment]
Closing this PR as stale. @ebozduman Please let me know if this needs to be kept opened.

**Pull request closed without merging**